tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+    const graphics::outline_color pipe_outline_color{2.0f, graphics::color{76, 47, 61}};
+    const float scroll_speed{200.f};
+    const std::string player_image_name{"player.png"};
+
+    // Background
+    const float ground_thickness{100.0f};
+    const float grass_thickness{20.0f};
+    const graphics::color background_color{82, 189, 199};
+    const graphics::color ground_color{220, 209, 143};
+    const graphics::color grass_color{132, 227, 90};
+    const graphics::outline_color grass_outline_color{2.0f, graphics::color{76, 47, 61}};
+};
+
+// Random number generator
+namespace {
+    std::random_device rd;  // Will be used to obtain a seed for the random number engine
+    std::mt19937 gen(rd()); // Standard mersenne_twister_engine seeded with rd()
+    float random_float(float lower, float higher) {
+        std::uniform_real_distribution<float> dist(lower, higher);
+        return dist(gen);
+    }
+}
+
+// A Flappy Bird column which has two pipes
+struct pipe {
+    entt::entity body{entt::null};
+    entt::entity cap{entt::null};
+
+    // Destroy pipe
+    void destroy(entt::registry &registry) {
+        registry.destroy(body);
+        registry.destroy(cap);
+    }
+};
+
+// Column is made of two pipes
+struct column {
+    // Entities representing the Flappy Bird pipes
+    pipe top_pipe{entt::null};
+    pipe bottom_pipe{entt::null};
+
+    // Is score taken from this column
+    bool scored{false};
+
+    // Destroy pipes and this column
+    void destroy(entt::registry &registry, entt::entity entity) {
+        top_pipe.destroy(registry);
+        bottom_pipe.destroy(registry);
+        registry.destroy(entity);
+    }
+};
+
+// Score struct, has current value, max record, and the UI text
+struct score {
+    int value;
+    int max_score;
+    entt::entity text;
+};
+
+// Logic functions
+namespace {
+    // Create the UI string
+    std::string score_ui_text(int score = 0, int best_score = 0) {
+        return std::string("Score: ") + std::to_string(score) +
+               std::string("\nBest: ") + std::to_string(best_score) +
+               std::string("\n\nW / UP / Space / Mouse to FLAP");
+    }
+
+    // Update score
+    void update_score(entt::registry &registry, entt::entity entity, bool reset = false) {
+        score &sc = registry.get<score>(entity);
+
+        // If reset is asked, set score to 0
+        if (reset) sc.value = 0;
+            // Else, increase the score,
+            // Compare it with the max score, and update max score if it's greater
+        else if (++sc.value > sc.max_score) sc.max_score = sc.value;
+
+        // Update the score entity
+        registry.assign_or_replace<score>(entity, sc);
+
+        // Update the UI text entity with the current values
+        auto &text = registry.get<graphics::text>(sc.text);
+        text.contents = score_ui_text(sc.value, sc.max_score);
+        registry.assign_or_replace<graphics::text>(sc.text, text);
```

Comment:
Here you are not forced to use `assign_or_replace`, `replace` is enough and do less check.